docroot/modules/custom/foia_quarterly_data_report/src/Plugin/Action/QuarterlyReportModerationAction.php

```diff
@@ -0,0 +1,296 @@
+      $plugin_definition,
+      $container->get('messenger'),
+      $container->get('content_moderation.moderation_information'),
+      $container->get('content_moderation.state_transition_validation')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function execute($entity = NULL) {
+    if (is_array($entity)) {
+      $this->executeMultiple($entity);
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function executeMultiple(array $entities) {
+    $batches = [];
+    foreach ($entities as $entity) {
+      $node = $this->getLatestRevision($entity);
+      $batches[] = [
+        '\Drupal\foia_quarterly_data_report\Plugin\Action\QuarterlyReportModerationAction::batchCore',
+        [$node, 'published', count($entities)],
+      ];
+    }
+
+    // Initialize batch (to set title).
+    $batch = [
+      'title' => t('Updating Moderation State'),
+      'operations' => $batches,
+      'finished' => '\Drupal\foia_quarterly_data_report\Plugin\Action\QuarterlyReportModerationAction::batchFinished',
+      'init_message' => t('Bulk operation of moderation state change is starting.'),
+      'progress_message' => t('Processed @current out of @total. Estimated time: @estimate.'),
+      'error_message' => t('The bulk operation process has encountered an error.'),
+    ];
+    batch_set($batch);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function access($object, AccountInterface $account = NULL, $return_as_object = FALSE) {
+    $user = !isset($account) ? \Drupal::service('current_user') : $account;
+    if ($object->getEntityTypeId() !== 'node') {
+      $this->messenger->addError($this->t('Can only perform publishing on FOIA quarterly report node content.'));
+      return FALSE;
+    };
+
+    if ($object->bundle() != 'quarterly_foia_report_data') {
+      $this->messenger->addError($this->t('Can only perform publishing on FOIA quarterly report content.'));
+      return FALSE;
+    }
+
+    if ($object->moderation_state->value != 'submitted_to_oip') {
+      $this->messenger->addError($this->t('Current moderation state is not a valide state for publishing.'));
+      return FALSE;
+    }
+
+    if (!$user->hasPermission('use quarterly report workflow transition publish')) {
+      $this->messenger->addError($this->t('User: %name does not have access to execute Moderation state change.', ['%name' => $user->getDisplayName()]));
+      return FALSE;
+    }
+
+    $field_agency = $object->get('field_agency')->getString();
+    $components = $object->get('field_agency_components')->getString();
+    if (empty($field_agency) || empty($components)) {
+      $this->messenger->addError($this->t('The agency and component field values must be validate for publishing.'));
+      return FALSE;
+    }
+
+    $workflows = $this->getTargetStates($object, $user);
+    $target_state = array_keys($workflows['target_states'])[0];
+    $valide_states = ['draft', 'published'];
+    if (!in_array($target_state, $valide_states)) {
```

Comment:
So I can understand better - what could a user do wrong that would trigger this? Why does it allow 'draft'?